teps/XXXX-step-workspaces.md

```diff
@@ -0,0 +1,254 @@
+      - name: my-workspace
+    ```
+
+The Sidecar `watch-workspace` will now have access to the `my-workspace` Workspace.
+
+### User Stories
+
+#### Story 1
+
+An author of the [`buildpacks-phases`](https://github.com/tektoncd/catalog/blob/master/task/buildpacks-phases/0.1/buildpacks-phases.yaml)
+Catalog task may want to rewrite the Task to reduce the possible blast radius of
+running untrusted images by limiting exposure of Docker credentials to only
+the Step which needs them to push images.
+
+In the buildpacks-phases Task there are 7 Steps and only 1 appears to need docker
+credentials. The are 6 other Steps that will currently receive creds-init docker
+credentials, run different images with different scripts and programs that could
+each be a vector to compromise those credentials.
+
+#### Story 2
+
+As a Task author I want to have a blessed method of sharing a Workspace between Steps and Sidecars
+so that I can pass configuration, messages, data, and logs between them.
+
+#### Story 3
+
+As a TaskRun and PipelineRun author I want to be able to trust that the Secret I provide
+via a Workspace Binding is only going to be available to code that is directly using it
+for its intended purpose.
+
+#### Story 4
+
+As a TaskRun and PipelineRun author I want to be able to quickly audit which
+containers, programs and scripts will have access to the credentials I am
+providing.
+
+### Risks and Mitigations
+
+## Design Details
+
+Add a new `WorkspaceUsage` struct to the `task_types.go` file:
+
+```go
+// WorkspaceUsage declares that a Step or Sidecar utilizes a Task's Workspace.
+type WorkspaceUsage struct {
+  // Name is the name of the Task's WorkspaceDeclaration that this Step or Sidecar is utilizing. It is required.
+  Name  string `json:"name"`
+}
+```
+
+Update the `Step` struct to include a slice of `WorkspaceUsage`:
+
+```go
+type Step struct {
+  corev1.Container `json:",inline"`
+  Script string `json:"script,omitempty"`
+
+  // Workspaces is a list of workspaces that this Step declares it will use in some way. The presence
+  // of a Workspace in this list prevents other Steps from automatically receiving the Workspace -
+  // they must also explicitly opt-in to receiving it as well in their own workspaces list.
+  Workspaces []WorkspaceUsage `json:"workspaces,omitempty"`
+}
+```
+
+Update the `Sidecar` struct to include a slice of `WorkspaceUsage`:
+
+```go
+// Sidecar embeds the Container type, which allows it to include fields not
+// provided by Container.
+type Sidecar struct {
+  corev1.Container `json:",inline"`
+  Script string `json:"script,omitempty"`
+
+  // Workspaces is a list of workspaces that this Sidecar declares it will use in some way.
+  Workspaces []WorkspaceUsage `json:"workspaces,omitempty"`
+}
+```
+
+## Drawbacks
+
+- It takes the `Step` type further from being a "pure" k8s Container. However we've already made moves
+in this direction by introducing our own `Script` field to `Steps`.
+
+## Alternatives
+
```

Comment:
another possible alternative, to tackle the sidecar issue specifically, would be to also make workspaces available to sidecars by default

Reply:
I mentioned this in another comment above but agreed that this could be a possible way forward but I'd prefer to have that convo in a new TEP rather than expand this one.  Happy to do that though if preferable.  For now I've added this as an explicit Non-Goal.

Reply:
Hmmm related question: why tackle these 2 features in the same TEP? Making workspaces available to sidecars and making steps be able to access them exclusively feel like different features to me - and maybe part of why I'm getting hung up about what feels to me like a mismatch in their functionality.

If we tackled them separately we could:

1. Have a TEP to add the ability to limit workspaces to only the steps that declare them
2. Then tackle making workspaces available to sidecars: we could decide if we want to mirror the step functionality, i.e. make workspaces automatically available to all sidecars, make it so that adding them explicitly to a sidecar excludes them from the rest; or not

Oh oh oh here is a question that I think might help explain my confusion about the "mismatch":

What if you want to add a workspace ONLY to a sidecar? e.g. the workspace contains something sensitive the sidecar needs, but you don't want to make it automatically available to all steps?

Reply:
> What if you want to add a workspace ONLY to a sidecar?

Great point and not something I'd considered _at all_.

OK, after our conversations here and online, a summary of where we're at:

1. Consider Steps and Sidecars as a single pool of containers
2. Sidecars automatically receive Workspaces, just as Steps do today
3. Adding a Workspace to a Step or Sidecar isolates it from _all other_ containers in the pool
    - If you add workspace to a Step you disable automatic mounting to other Steps & Sidecars
    - If you add workspace to a Sidecar you prevent automatic mounting to other Sidecars & Steps